### PR TITLE
feat(autoquality): walk :ssim2 search to target (symmetric allowed_error)

### LINKED ISSUE #366: autoquality: switch :ssim2 search to walk-to-target (symmetric allowed_error) semantics
## Summary

The `:ssim2` autoquality search currently treats `allowed_error` as a **one-sided floor and walks to the *lowest* satisfying quality** (byte-minimizing). The intended design was a **symmetric tolerance band around `target`** with the search **converging toward `target`** (walk-to-target) — aim for `target`, accept anything in `[target − allowed_error, target + allowed_error]`, stop when close, don't keep descending. This issue tracks switching to the intended semantics.

This is a deliberate, observable behavior change to the search; it is **not** an imgproxy-compat break (our `:ssim2` `allowed_error` is already documented as *not* imgproxy's DSSIM `allowed_error`), and the project is pre-release greenfield, so now is the cheap time to fix it.

## Current behavior

[`ImagePipe.Output.EncodeSearch.objective_phase/3`](lib/image_pipe/output/encode_search.ex) for `:ssim2`:

```elixir
band = rqs.target - rqs.allowed_error            # 78 - 1 = 77
predicate = fn _bytes, score -> score >= band end
search_lowest_satisfying(rqs.min_quality, rqs.max_quality, predicate, ctx)
```

- The band is `[target − allowed_error, ∞)` — the `target + allowed_error` **upper edge is never referenced**.
- Among all qualities scoring `≥ floor`, it ships the **smallest file** (lowest q), i.e. it actively walks *down toward the floor*, not toward `target`.
- Net: with `target 78 / allowed_error 1`, an image is shipped at the **bottom** of the intended band (~77.x), systematically *under*-delivering the configured quality.

Documented as-is in [`ImagePipe.Plan.Output.QualitySearch`](lib/image_pipe/plan/output/quality_search.ex) ("accepts the lowest quality scoring ≥ target − allowed_error").

## Proposed behavior (walk-to-target)

- Acceptance band is **symmetric**: `[target − allowed_error, target + allowed_error]`.
- The search **converges toward `target`** (binary search; in-band → accept and stop; overshoot → go lower; undershoot → go higher), rather than minimizing quality below the floor.
- `allowed_error` becomes an intuitive *stopping tolerance around the target*, not a downward-only quality sacrifice.

This is the behavior prototyped as the **"two-sided"** variant in `mix autoquality.bench --part g`.

## Evidence (autoquality bench, `docs/autoquality_benchmark.md`)

The bench makes the case that the byte-saving rationale for walk-to-floor is weak, so the change is nearly free while delivering the configured quality more honestly. Pinned codec-corpus, Apple Silicon / libvips 8.18.2 (portable conclusion is the shape, not exact numbers):

**Part G — two-sided (≈ walk-to-target) vs current lowest-satisfying:**

| format | under-target: current → walk-to-target | byte cost |
|---|---|---|
| jpeg | 74% → 66% | +0.6% |
| webp | 89% → 87% | +0.3% |
| avif | 56% → 44% | +0.8% |

→ meaningfully more on-target for **+0.3–0.8%** bytes — below the "worth optimizing" threshold. The big autoquality savings (easy images hitting target at a low quality) survive, because both behaviors ship those at the lowest q that *reaches* target; walk-to-floor only squeezes the `[target−ae, target)` sliver.

**Part J — sweeping `allowed_error` (current logic) confirms the same shape from the other direction:** tightening `allowed_error` 1→0 buys **+15–21 pts on-target for ~1–2.5 kB** (jpeg 20→39%, webp 11→25%, avif 41→63%), because the floor-walking leaves easy on-target quality on the table.

## Design points to settle (before code)

1. **Empty-band fallback (integer-quality granularity).** A symmetric `[77, 79]` band can be *empty* for an image (e.g. q73 → 76.8, q74 → 79.5, nothing in band). Define the fallback: nearest-overshoot (ship the just-above q) vs nearest in either direction, and the ceiling/floor behavior when `target` is unreachable. Part G's two-sided used overshoot-then-ceiling; confirm that's the intended rule.
2. **Crop confirm/bump alignment.** The crop confirm band also derives from `target − allowed_error`, and the bump phase already walks *up* toward the band — so today the objective walks *down* while the confirm walks *up* (internally inconsistent). Walk-to-target makes the pipeline coherent; reconcile `confirm_band` / bump direction in the same change.
3. **Docs + conformance.** Rewrite the `QualitySearch` moduledoc, update `docs/imgproxy_support_matrix.md` (behavioral axis — note the deliberate divergence), and refresh `docs/autoquality_benchmark.md` finding #9 to describe the shipped semantics rather than the floor-walking one.
4. **Tests.** Request-boundary test decoding the response and asserting delivered score tracks `target` (not the floor); EncodeSearch unit tests for in-band/overshoot/undershoot/empty-band; cache-key unchanged.

## Extra recommendations surfaced by the same data

In scope for this change:

- **Revisit the default `allowed_error` under the new semantics.** Once `allowed_error` is a symmetric tolerance, `1.0` means "±1 around target" — reasonable, but reconfirm it's the right default (Parts G/J are the harness).

Related, data-driven, **better as their own issues** (captured here so they aren't lost):

- **Lower the crop-scoring crossover below 6 MP.** Part H validated crop+confirm against full-frame on the >6 MP cohort with **zero target-hit regressions** (worst delivered undershoot −0.24 pt), de-risking capturing the size-independent per-probe metric cost on the 2–6 MP band (most real delivery sizes). Needs validation on more >6 MP content. (Distinct from #359, which sweeps K/tile size.)
- **Ship a non-zero `autoquality_max_resolution` default.** Still defaults to `0` (unbounded); Part A shows the search is multi-second on large images. The deadline (#355) + crop-scoring eventually make it a policy knob, but an interim cap is the safe latency guard.
- **Document the on-target lever hierarchy for hosts:** `allowed_error` (cheap, broad on-target lift) → `max_quality` (per-format top-up for ceiling-pinned content). Both host policy, not defaults changes.

## Scope / files

- `lib/image_pipe/output/encode_search.ex` (objective phase predicate + search direction; crop confirm band/bump)
- `lib/image_pipe/plan/output/quality_search.ex` (moduledoc)
- `docs/imgproxy_support_matrix.md`, `docs/autoquality_benchmark.md`
- tests at the request boundary + EncodeSearch unit

### PR DESCRIPTION
**Pull request by hlindset**
## Summary

Switches the `:ssim2` autoquality search from byte-minimizing **walk-to-floor** to **walk-to-target** semantics, per #366.

The old search treated `allowed_error` as a one-sided floor and walked to the *lowest* satisfying quality — so with `allowed_error 1` it shipped boundary images at ~`target − 1`, systematically under-delivering the configured quality. The search now converges on the **symmetric** band `[target − allowed_error, target + allowed_error]`, accepting the first in-band quality and stopping.

This is a deliberate, observable behavior change (not an imgproxy-compat break — our `:ssim2` `allowed_error` is already documented as *not* imgproxy's DSSIM `allowed_error`), and the project is pre-release greenfield.

## Design decisions (confirmed up front)

1. **Empty band** (integer-quality granularity straddles it) → ship the nearest **overshoot** (the just-above quality, a `:hit`, never under-delivers); unreachable target → pin to `max_quality`, `:best_effort`/`:ceiling`. (Part G's two-sided prototype rule.)
2. **Crop confirm** kept as a one-sided floor (`target − allowed_error`) with upward bump — now coherent with the objective (objective aims for the center, confirm guards the floor).
3. **Default `allowed_error = 1.0` retained** — now an intuitive ±1 tolerance around the target rather than a downward sacrifice.

## Changes

- `lib/image_pipe/output/encode_search.ex` — `search_to_target/5` (in-band → accept; overshoot → lower; undershoot → higher); removed the dead lowest-satisfying primitives; added `probe_score/2`; updated crop-offset comments.
- `lib/image_pipe/plan/output/quality_search.ex`, `lib/image_pipe/parser/imgproxy.ex` — moduledoc/comment rewritten for symmetric semantics.
- `docs/imgproxy_support_matrix.md` — behavioral-axis update + deliberate-divergence note.
- `docs/autoquality_benchmark.md` — finding #9 refreshed to the shipped semantics; reconciled the now-contradicting present-tense claims in #6, #8, and the bottom line.

## Tests

- **Unit** (`encode_search_test.exs`): converge-to-target-not-floor, empty-band overshoot, easy-image floor overshoot, ceiling undershoot, crop-mode reconciliation.
- **Wire** (`imgproxy_wire_conformance_test.exs`): scores the response against a lossless-PNG reference (the search's exact metric) and asserts the delivered score tracks the target (≈82.7 with band `[77,83]`), not the floor.
- Property invariant (`score ≥ target − allowed_error` for hits) unchanged and still holds. Cache key unchanged.

Gate: `mix format --check-formatted`, `compile --warnings-as-errors`, `credo --strict`, full `mix test` (2545 tests + 81 properties + 1 doctest, 0 failures) all pass.

## Out of scope (per #366 — better as their own issues)

Lower the crop-scoring crossover below 6 MP; ship a non-zero `autoquality_max_resolution` default; host-facing lever-hierarchy doc.

Fixes #366

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Quality search now uses a symmetric tolerance band around the target, accepting the first quality within the band rather than minimizing to the lower bound. The `allowed_error` parameter now correctly controls the band width.

* **Documentation**
  * Updated parameter documentation and benchmark guidance to clarify symmetric tolerance band behavior and walk-to-target semantics.

* **Tests**
  * Added test coverage for walk-to-target behavior and edge cases (zero-width bands, empty bands, unreachable targets).

### CODE REVIEW

File: lib/image_pipe/parser/imgproxy.ex [hunk 1]
```diff
@@ -59,9 +59,10 @@ defmodule ImagePipe.Parser.Imgproxy do
...
                      autoquality_min_quality: [type: :pos_integer, default: 70],
                      autoquality_max_quality: [type: :pos_integer, default: 80],
-                     # SSIMULACRA2-scale band (points on 0–100), NOT imgproxy's DSSIM
-                     # `allowed_error` (0–1, 1.0 = accept anything); here 1.0 is a strict
-                     # one-point band. Applies only to `:ssim2`. See QualitySearch.
+                     # Symmetric SSIMULACRA2-scale tolerance band (points on 0–100),
+                     # NOT imgproxy's DSSIM `allowed_error` (0–1, 1.0 = accept anything);
+                     # here 1.0 is a strict ±1-point band around the target. Applies only
+                     # to `:ssim2`. See QualitySearch.
                      autoquality_allowed_error: [type: {:or, [:integer, :float]}, default: 1.0],
```
[the remark this answers is not in the document]
Comment (hlindset) [reply lib/image_pipe/parser/imgproxy.ex]: Fixed in f4c1d646. The URL form was already guarded by the grammar's `parse_non_negative_float` (option_grammar.ex:363), so the gap was host config only; added a `:custom` non-negative validator on `autoquality_allowed_error`, consistent with the existing `validate_autoquality_brackets!` config-sanity checks, plus a config-boundary rejection test.
